src/GT/DB/Component.hs

```diff
@@ -21,3 +23,22 @@ mileage componentId =
     on $ usage ^. UsageRide ==. ride ^. RideId
     where_ $ usage ^. UsageComponent ==. val componentId
     pure $ sum_ $ ride ^. RideDistance
+
+getForBike :: MonadIO m => BikeId -> SqlPersistT m [Entity Component]
+getForBike bikeId =
+    [sqlQQ|
+WITH RECURSIVE children AS (
+    SELECT c.*
+    FROM component AS c
+    WHERE c.id = #{bikeId}
+
+    UNION
+
+    SELECT c_r.*
+    FROM component AS c_r
+    INNER JOIN children AS c
+        ON c.id = c_r.parent
+    )
+SELECT ??
+FROM children AS component
```

Comment:
Is the `factored` referring to the use of raw sql here?

Reply:
More or less I was interested in why you are using a sql quasi quoter here.

Reply:
esqueleto does not support common table expressions or `UNION` so I needed to drop to raw SQL here.

Beforehand, the function `Usage.new` looked like:

```haskell
new bikeId rideId = do
  componentIds <- [rawQQ|
the
whole
big
query
  |]
  for componentIds $ \componentId -> do
    insert $ Usage bikeId componentId
```

In the test I wrote for `Usage.new`, i asserted that ``length usages `shouldBe` 2``, and this was indeed the answer I got back. However, that was because the underlying `componentIds` query was incorrect. I separated the SQL query into it's own function, and had it return `component`s instead, and then I could write a good test that verified that it did what I wanted it to.

This function then becomes a somewhat trivial composition of functions and doesn't really need a test, but I kept it anyway.

Reply:
Ah right, thanks!

Reply:
Beforehand, the query lived in `Usage.new`. I factored it out to `Component.getForBike`, so that I could test it independently of `Usage.new`.

I'm using a raw SQL here because esqueleto doesn't support UNION or recursive CTEs.